core/vcpu.c

```diff
@@ -2581,30 +2581,27 @@ static void handle_cpuid_virtual(struct vcpu_t *vcpu, uint32_t a, uint32_t c)
             FEATURE(PSE)        |
             FEATURE(HTT);
 
-    static uint32_t cpu_features_2 =
+    static uint32_t cpuid_1_features_ecx =
             FEATURE(SSE3)       |
             FEATURE(SSSE3)      |
             FEATURE(SSE41)      |
             FEATURE(SSE42)      |
             FEATURE(CMPXCHG16B) |
             FEATURE(MOVBE)      |
+            FEATURE(AESNI)      |
+            FEATURE(PCLMULQDQ)  |
             FEATURE(POPCNT);
 
-    uint32_t cpu_features_ext =
+    static uint32_t cpuid_8000_0001_features_edx =
             FEATURE(NX)         |
             FEATURE(SYSCALL)    |
+            FEATURE(RDTSCP)     |
+            FEATURE(PDPE1GB)    |
             FEATURE(EM64T);
 
-    // Conditional features
-    if (cpu_has_feature(X86_FEATURE_AESNI)) {
-        cpu_features_2 |= FEATURE(AESNI);
-    }
-    if (cpu_has_feature(X86_FEATURE_PCLMULQDQ)) {
-        cpu_features_2 |= FEATURE(PCLMULQDQ);
-    }
-    if (cpu_has_feature(X86_FEATURE_RDTSCP)) {
-        cpu_features_ext |= FEATURE(RDTSCP);
-    }
```

Comment:
The original code here is to verify host feature before report to guest. Although modern CPU support them, we may need some logic to verify them before report to guest. On one hand we can filter out host unsupported feature since haxm behaves CPU features as subset of the host CPU, on the other in future we can disable certain features reported to guest for specific purpose like customize CPU feature.

So add a function here to verify 80000001 ecx/edx before the switch case.

Reply:
handle_cpuid_virtual returns the intersection of supported by the real processor features (state->_eax etc) and supported by the hypervisor. 
cpu_has_feature calls are redundant, since not supported features are not in state->_eax etc.
Hax doesn't support cpu features setting from qemu, so these lists are hardcoded. When such support will be added, hax will get these feature lists from qemu.

Reply:
Actually we have plan on the list to add the support of setting cpu feature from qemu, so I'm thinking adding such validation for a CPUID handling refactoring patch is suitable.

Reply:
The current list of supported hax cpuid features is nearby of nehalem processor, which is very old. Getting feature list from qemu has any meaning when you support the newest CPUs and allow users to run hax with older CPU features. To support newer processors a lot of features should be added to hax, not just returning feature bits for cpuid. Some of them can slow down it's execution by generating new vm-exits (example - xsave support). When I tried to support Win10, I tried to add more features and experienced this.
Even if you add them, why do you want to check them in cpuid handling code, and not in ioctl? I think when the user wants to run a specific feature set and the underlying CPU does not support some of them, it would be correct to check this in ioctl handling code and report an error.
Even if you want to check them in cpuid code, why should it be done with a specific function - handle_cpuid_virtual already has the supported by the CPU features in it's state (eax\ebx\ecx\edx) and a simple intersection with the supported (by example state->_ecx = cpuid_8000_0001_features_ecx & state->_ecx) does the same work as checking the features and adding them to the list (example - cpuid_8000_0001_features_ecx).

Reply:
Maybe I made this complicated a little bit early.

Yes currently all CPUID handled by handle_cpuid_virtual() will first return from host asm_cpuid() then do the filtering, so check host has the feature is redundant.

The case I'm describing is that, if due to some reason QEMU wants to test some feature that hasn't been added to haxm, via ioctl, like recent PCLMULQDQ. Current logic will clear them. What I'm expecting is some function which can validate those features, it doesn't have to be in handle_cpuid_virtual(), could be near iotcl as you said. 
You can leave this to us, we'll make change together when we add the interface.